File: frontend/src/types/presupuesto.ts
// ============================================================
// Tipos TypeScript — Módulo Presupuesto
// Nomenclatura en español conforme a Regla Arquitectónica 2
// ============================================================

// ── PartidaPresupuestaria ────────────────────────────────────
export type TipoPartida = 'INGRESO' | 'GASTO';

export interface PartidaPresupuestaria {
  id_partida: number;
  codigo: string;
  nombre: string;
  descripcion?: string;
  tipo: TipoPartida;
  nivel: number;
  es_hoja: boolean;
  id_partida_padre?: number;
  estado: string;
  creado_en: string;
}

export interface PartidaPresupuestariaCrear {
  codigo: string;
  nombre: string;
  descripcion?: string;
  tipo: TipoPartida;
  nivel: number;
  es_hoja?: boolean;
  id_partida_padre?: number;
  estado?: string;
}

export interface PartidaPresupuestariaActualizar {
  nombre?: string;
  descripcion?: string;
  es_hoja?: boolean;
  estado?: string;
}

// ── PresupuestoAnual ─────────────────────────────────────────
export interface PresupuestoAnual {
  id_presupuesto: number;
  anio_fiscal: number;
  denominacion: string;
  monto_inicial: number;
  monto_codificado: number;
  estado: string;
  fecha_aprobacion?: string;
  resolucion_aprobacion?: string;
  observaciones?: string;
  creado_en: string;
}

export interface PresupuestoAnualCrear {
  anio_fiscal: number;
  denominacion: string;
  monto_inicial?: number;
  observaciones?: string;
}

// ── AsignacionPresupuestaria ─────────────────────────────────
export interface AsignacionPresupuestaria {
  id_asignacion: number;
  id_presupuesto: number;
  id_partida: number;
  monto_inicial: number;
  monto_codificado: number;
  monto_comprometido: number;
  monto_devengado: number;
  monto_pagado: number;
  saldo_disponible?: number;
  estado: string;
  creado_en: string;
}

export interface AsignacionPresupuestariaCrear {
  id_presupuesto: number;
  id_partida: number;
  monto_inicial?: number;
}

// ── ReformaPresupuestaria ────────────────────────────────────
export type TipoReforma = 'TRASPASO' | 'SUPLEMENTO' | 'REDUCCION';

export interface ReformaPresupuestaria {
  id_reforma: number;
  id_asignacion: number;
  tipo_reforma: TipoReforma;
  monto: number;
  numero_resolucion: string;
  fecha_resolucion: string;
  motivo?: string;
  estado: string;
  creado_en: string;
}

export interface ReformaPresupuestariaCrear {
  id_asignacion: number;
  tipo_reforma: TipoReforma;
  monto: number;
  numero_resolucion: string;
  fecha_resolucion: string;
  motivo?: string;
}

// ── CertificadoPresupuestario ────────────────────────────────
export type EstadoCertificado = 'PENDIENTE' | 'APROBADO' | 'ANULADO' | 'LIQUIDADO';

export interface CertificadoPresupuestario {
  id_certificado: number;
  id_asignacion: number;
  numero_certificado: string;
  monto_certificado: number;
  concepto: string;
  fecha_solicitud: string;
  fecha_certificacion?: string;
  fecha_vencimiento?: string;
  referencia_tipo?: string;
  referencia_id?: number;
  id_proceso_contratacion?: number;
  estado: EstadoCertificado;
  motivo_anulacion?: string;
  creado_en: string;
}

export interface CertificadoPresupuestarioCrear {
  id_asignacion: number;
  monto_certificado: number;
  concepto: string;
  fecha_solicitud: string;
  fecha_vencimiento?: string;
  referencia_tipo?: string;
  referencia_id?: number;
  id_proceso_contratacion?: number;
}

export interface CertificadoPresupuestarioEstado {
  estado: 'APROBADO' | 'ANULADO';
  motivo_anulacion?: string;
}

// ── Compromiso ───────────────────────────────────────────────
export interface Compromiso {
  id_compromiso: number;
  id_certificado: number;
  numero_compromiso: string;
  monto_comprometido: number;
  concepto: string;
  fecha_compromiso: string;
  estado: string;
  motivo_anulacion?: string;
  creado_en: string;
}

export interface CompromisoCrear {
  id_certificado: number;
  numero_compromiso: string;
  monto_comprometido: number;
  concepto: string;
  fecha_compromiso: string;
}

// ── Devengado ────────────────────────────────────────────────
export interface Devengado {
  id_devengado: number;
  id_compromiso: number;
  numero_devengado: string;
  monto_devengado: number;
  concepto: string;
  fecha_devengado: string;
  id_asiento_contable?: number;
  id_factura?: number;
  estado: string;
  motivo_anulacion?: string;
  creado_en: string;
}

export interface DevengadoCrear {
  id_compromiso: number;
  numero_devengado: string;
  monto_devengado: number;
  concepto: string;
  fecha_devengado: string;
  id_asiento_contable?: number;
  id_factura?: number;
}

// ── EjecucionPresupuestaria ──────────────────────────────────
export interface EjecucionPresupuestaria {
  anio_fiscal: number;
  codigo_partida: string;
  nombre_partida: string;
  monto_inicial: number;
  monto_codificado: number;
  monto_comprometido: number;
  monto_devengado: number;
  monto_pagado: number;
  saldo_disponible: number;
  porcentaje_ejecucion: number;
}

File: backend/app/schemas/presupuesto.py
"""
Schemas Pydantic — Módulo Presupuesto
Nomenclatura en español (Regla 2 obligatoria).
"""
from __future__ import annotations
from datetime import date, datetime
from decimal import Decimal
from typing import List, Optional
from pydantic import BaseModel, Field, model_validator


# ===========================================================================
# PARTIDA PRESUPUESTARIA
# ===========================================================================

class PartidaPresupuestariaBase(BaseModel):
    codigo: str = Field(..., max_length=30, description="Código SIGEF ej: 5.1.01.05.001")
    nombre: str = Field(..., max_length=255)
    descripcion: Optional[str] = None
    tipo: str = Field(..., pattern="^(INGRESO|GASTO)$")
    nivel: int = Field(..., ge=1, le=5)
    es_hoja: bool = True
    id_partida_padre: Optional[int] = None
    estado: str = "ACTIVO"


class PartidaPresupuestariaCrear(PartidaPresupuestariaBase):
    pass


class PartidaPresupuestariaActualizar(BaseModel):
    nombre: Optional[str] = None
    descripcion: Optional[str] = None
    es_hoja: Optional[bool] = None
    estado: Optional[str] = None


class PartidaPresupuestariaRespuesta(PartidaPresupuestariaBase):
    id_partida: int
    creado_en: datetime

    class Config:
        from_attributes = True


# ===========================================================================
# PRESUPUESTO ANUAL
# ===========================================================================

class PresupuestoAnualBase(BaseModel):
    anio_fiscal: int = Field(..., ge=2000, le=2100)
    denominacion: str = Field(..., max_length=255)
    monto_inicial: Decimal = Field(default=Decimal("0.00"), ge=0)
    observaciones: Optional[str] = None


class PresupuestoAnualCrear(PresupuestoAnualBase):
    pass


class PresupuestoAnualActualizar(BaseModel):
    denominacion: Optional[str] = None
    monto_codificado: Optional[Decimal] = None
    estado: Optional[str] = None
    fecha_aprobacion: Optional[date] = None
    resolucion_aprobacion: Optional[str] = None
    observaciones: Optional[str] = None


class PresupuestoAnualRespuesta(PresupuestoAnualBase):
    id_presupuesto: int
    monto_codificado: Decimal
    estado: str
    fecha_aprobacion: Optional[date]
    resolucion_aprobacion: Optional[str]
    creado_en: datetime

    class Config:
        from_attributes = True


# ===========================================================================
# ASIGNACIÓN PRESUPUESTARIA
# ===========================================================================

class AsignacionPresupuestariaBase(BaseModel):
    id_presupuesto: int
    id_partida: int
    monto_inicial: Decimal = Field(default=Decimal("0.00"), ge=0)


class AsignacionPresupuestariaCrear(AsignacionPresupuestariaBase):
    pass


class AsignacionPresupuestariaRespuesta(AsignacionPresupuestariaBase):
    id_asignacion: int
    monto_codificado: Decimal
    monto_comprometido: Decimal
    monto_devengado: Decimal
    monto_pagado: Decimal
    saldo_disponible: Optional[Decimal] = None
    estado: str
    creado_en: datetime

    class Config:
        from_attributes = True


# ===========================================================================
# REFORMA PRESUPUESTARIA
# ===========================================================================

class ReformaPresupuestariaCrear(BaseModel):
    id_asignacion: int
    tipo_reforma: str = Field(..., pattern="^(TRASPASO|SUPLEMENTO|REDUCCION)$")
    monto: Decimal = Field(..., description="Positivo=aumento, negativo=reducción")
    numero_resolucion: str = Field(..., max_length=100)
    fecha_resolucion: date
    motivo: Optional[str] = None


class ReformaPresupuestariaRespuesta(ReformaPresupuestariaCrear):
    id_reforma: int
    estado: str
    creado_en: datetime

    class Config:
        from_attributes = True


# ===========================================================================
# CERTIFICADO PRESUPUESTARIO
# ===========================================================================

class CertificadoPresupuestariaBase(BaseModel):
    id_asignacion: int
    monto_certificado: Decimal = Field(..., gt=0)
    concepto: str
    fecha_solicitud: date
    fecha_vencimiento: Optional[date] = None
    referencia_tipo: Optional[str] = None
    referencia_id: Optional[int] = None
    id_proceso_contratacion: Optional[int] = None


class CertificadoPresupuestariaCrear(CertificadoPresupuestariaBase):
    pass


class CertificadoPresupuestariaEstado(BaseModel):
    estado: str = Field(..., pattern="^(APROBADO|ANULADO)$")
    motivo_anulacion: Optional[str] = None

    @model_validator(mode="after")
    def validar_motivo_anulacion(self) -> "CertificadoPresupuestariaEstado":
        if self.estado == "ANULADO" and not self.motivo_anulacion:
            raise ValueError("motivo_anulacion es obligatorio al anular")
        return self


class CertificadoPresupuestariaRespuesta(CertificadoPresupuestariaBase):
    id_certificado: int
    numero_certificado: str
    fecha_certificacion: Optional[date]
    estado: str
    motivo_anulacion: Optional[str]
    creado_en: datetime

    class Config:
        from_attributes = True


# ===========================================================================
# COMPROMISO
# ===========================================================================

class CompromisoCrear(BaseModel):
    id_certificado: int
    numero_compromiso: str = Field(..., max_length=50)
    monto_comprometido: Decimal = Field(..., gt=0)
    concepto: str
    fecha_compromiso: date


class CompromisoRespuesta(CompromisoCrear):
    id_compromiso: int
    estado: str
    motivo_anulacion: Optional[str]
    creado_en: datetime

    class Config:
        from_attributes = True


# ===========================================================================
# DEVENGADO
# ===========================================================================

class DevengadoCrear(BaseModel):
    id_compromiso: int
    numero_devengado: str = Field(..., max_length=50)
    monto_devengado: Decimal = Field(..., gt=0)
    concepto: str
    fecha_devengado: date
    id_asiento_contable: Optional[int] = None
    id_factura: Optional[int] = None


class DevengadoRespuesta(DevengadoCrear):
    id_devengado: int
    estado: str
    motivo_anulacion: Optional[str]
    creado_en: datetime

    class Config:
        from_attributes = True


# ===========================================================================
# RESUMEN EJECUCIÓN PRESUPUESTARIA
# ===========================================================================

class EjecucionPresupuestariaRespuesta(BaseModel):
    anio_fiscal: int
    codigo_partida: str
    nombre_partida: str
    monto_inicial: Decimal
    monto_codificado: Decimal
    monto_comprometido: Decimal
    monto_devengado: Decimal
    monto_pagado: Decimal
    saldo_disponible: Decimal
    porcentaje_ejecucion: Decimal

    class Config:
        from_attributes = True
